Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -876,7 +861,7 @@ extension SwiftLanguageServer {
         return
       }
 
-      let tokens = snapshot.tokens.mergedAndSorted
+      let tokens = snapshot.tokens.semantic.sorted(by: { $0.start < $1.start })
```

Comment:
How difficult would it be to redo the syntactic highlighting with the syntax tree? Specifically, it's not clear to me where SourceKit respects the "syntactic only" flag when generating the syntactic token stream in the first place.

Reply:
> IIRC merging the syntactic and semantic highlighting provided from sourcekitd was essential to provide good semantic highlighting because sourcekitd’s semantic highlights assume that the syntactic highlights shine through.

This is true, although we also have some level of lexical/syntactic highlighting on the VSCode side with its Swift grammar already (this is IIRC the original motivation for calling the LSP feature 'semantic highlighting', so only the heavy lifting is delegated to the language server). Still, other LSP clients might not have a separate grammar, additionally the quality of sourcekitd's syntactic tokens is (in my experience) considerably better than VSCode's TextMate grammar, so I'd be in favor of keeping lexical/syntactic highlighting if possible.

> How difficult would it be to redo the syntactic highlighting with the syntax tree?

Doing the syntactic highlighting with swift-syntax would probably be quite doable if the plan is to maintain a syntax tree on the language server side anyway. Is the long-term plan to let sourcekitd only handle the semantic tokens?

Reply:
@CodaFi and I discussed this offline yesterday and I think we should be able to use SwiftSyntax’s `SyntaxClassifier` for syntactic highlighting (in my memory that one was even superior to `sourcekitd`’s syntactic highlighting) and only use `sourcekitd` for semantic highlighting.